src/server/client.ts

```diff
@@ -0,0 +1,505 @@
+            this.host.writeMessage(message);
+        }
+
+        private getLineMap(fileName: string): number[] { 
+            var lineMap = ts.lookUp(this.lineMaps, fileName);
+            if (!lineMap) {
+                var scriptSnapshot = this.host.getScriptSnapshot(fileName);
+                lineMap  = this.lineMaps[fileName] = ts.computeLineStarts(scriptSnapshot.getText(0, scriptSnapshot.getLength()));
+            }
+            return lineMap;
+        }
+
+        private lineColToPosition(fileName: string, lineCol: ts.server.protocol.LineCol): number {
+            return ts.computePositionFromLineAndCharacter(this.getLineMap(fileName), lineCol.line, lineCol.col);
+        }
+
+        private positionToOneBasedLineCol(fileName: string, position: number): ts.server.protocol.LineCol {
+            var lineCol = ts.computeLineAndCharacterOfPosition(this.getLineMap(fileName), position);
+            return {
+                line: lineCol.line,
+                col: lineCol.character
+            };
+        }
+
+        private convertCodeEditsToTextChange(fileName: string, codeEdit: ts.server.protocol.CodeEdit): ts.TextChange {
+            var start = this.lineColToPosition(fileName, codeEdit.start);
+            var end = this.lineColToPosition(fileName, codeEdit.end);
+
+            return {
+                span: ts.createTextSpanFromBounds(start, end),
+                newText: codeEdit.newText
+            };
+        }
+
+        private decodeEncodedFileId(fileId: ts.server.protocol.EncodedFile): string {
+            var fileName: string;
+            if (typeof fileId === "object") {
+                fileName = (<ts.server.protocol.IdFile>fileId).file;
+                this.fileMapping[(<ts.server.protocol.IdFile>fileId).id] = fileName;
+            }
+            else if (typeof fileId === "number") {
+                fileName = ts.lookUp(this.fileMapping, fileId.toString());
+                Debug.assert(!!fileName, "Did not find filename in previous fileID mappings.");
+            }
+            else {
+                Debug.fail("Got unexpedted fileId type.");
+            }
+            return fileName;
+        }
+
+        private processRequest<T extends ts.server.protocol.Request>(command: string, arguments?: any): T {
+            var request: ts.server.protocol.Request = {
+                seq: this.sequence++,
+                type: "request",
+                command: command,
+                arguments: arguments
+            };
+
+            this.writeMessage(JSON.stringify(request));
+
+            return <T>request;
+        }
+
+        private processResponse<T extends ts.server.protocol.Response>(request: ts.server.protocol.Request): T {
+            var lastMessage = this.messages.shift();
+            Debug.assert(!!lastMessage, "Did not recieve any responses.");
+
+            // Read the content length
+            var contentLengthPrefix = "Content-Length: ";

```

Comment:
Why are we decomposing http headers here?  Why not just use a real http library for this?